vm/ubpf_jit_x86_64.h

```diff
@@ -637,4 +637,246 @@ emit_dispatched_external_helper_call(struct jit_state* state, unsigned int idx)
+ * @param[in] offset Offset from the destination register
+ */
+static inline void
+emit_atomic_alu(struct jit_state* state, int opcode, int is_64bit, int src, int dst, int offset)
+{
+    emit1(state, 0xf0); // lock prefix
+    emit_basic_rex(state, is_64bit, src, dst);
+    emit1(state, opcode);
+    emit_modrm_and_displacement(state, src, dst, offset);
+}
+
+/**
+ * @brief Atomically compare and exchange.
+ * This writes the value stored in src into the memory location identified by dst and offset, if the value at that
+ * location is equal to the value in RAX.
+ *
+ * @param[in,out] state The JIT state
+ * @param[in] is_64bit Whether the operation is 64-bit or 32-bit
+ * @param[in] src The value to write.
+ * @param[in] dst The base address of the destination memory location.
+ * @param[in] offset The offset from the destination memory location.
+ * @note This implicitly uses RAX as the original value to compare against and stores the original value in [destination
+ * + offset] into RAX.
+ */
+static inline void
+emit_atomic_compare_exchange(struct jit_state* state, int is_64bit, int src, int dst, int offset)
+{
+    emit1(state, 0xf0); // lock prefix
+    emit_basic_rex(state, is_64bit, src, dst);
+    emit1(state, 0x0f);
+    emit1(state, 0xb1);
+    emit_modrm_and_displacement(state, src, dst, offset);
+}
+
+/**
+ * @brief Atomically write the value of src into the memory location identified by dst and offset.
+ *
+ * @param[in,out] state The JIT state
+ * @param[in] is_64bit Whether the operation is 64-bit or 32-bit
+ * @param[in] src The value to write.
+ * @param[in] dst The base address of the destination memory location.
+ * @param[in] offset The offset from the destination memory location.
+ * @note This operation stores the original value in [destination + offset] into RAX.
+ */
+static inline void
+emit_atomic_exchange(struct jit_state* state, int is_64bit, int src, int dst, int offset)
+{
+    emit1(state, 0xf0); // lock prefix
+    emit_basic_rex(state, is_64bit, src, dst);
+    emit1(state, 0x87);
+    emit_modrm_and_displacement(state, src, dst, offset);
+}
+
+/**
+ * @brief Perform an ALU operation atomically and fetch the result.
+ *
+ * @param[in,out] state The JIT state
+ * @param[in] is_64bit Whether the operation is 64-bit or 32-bit
+ * @param[in] opcode The x64 opcode for the ALU operation
+ * @param[in] src The source register, used as the operand for the ALU operation.
+ * @param[in] dst The base address of the destination memory location.
+ * @param[in] offset The offset from the destination memory location.
+ * @note This operation stores the original value in [destination + offset] into RAX.
+ * @note This operation is emulated on x64 for 64-bit and 32-bit operations as the x64 architecture does not have an
+ * atomic fetch-and, fetch-or, fetch-xor, or fetch-add instruction.
+ */
+static inline void
+emit_atomic_fetch_alu(struct jit_state* state, int is_64bit, int opcode, int src, int dst, int offset)
```

Comment:
FYI: We reviewed this function synchronously on slack.

Reply:
Was there any proposed changes? I don't remember.